File: hlsf_db_tools/adjacency_families.py
"""Shared adjacency family classification used by exporters and tooling."""

from __future__ import annotations

from enum import Enum


class AdjacencyFamily(str, Enum):
    SPATIAL = "spatial"
    TEMPORAL = "temporal"
    CAUSAL = "causal"
    HIERARCHICAL = "hierarchical"
    ANALOGICAL = "analogical"
    CONSTRAINT = "constraint"
    VALUE = "value"
    COMMUNICATIVE = "communicative"
    SOCIAL = "social"
    MODAL = "modal"
    EVIDENTIAL = "evidential"
    COUNTERFACTUAL = "counterfactual"
    OPERATIONAL = "operational"
    MEASUREMENT = "measurement"
    AESTHETIC = "aesthetic"


RELATION_FAMILY_MAP: dict[str, AdjacencyFamily] = {
    "proximity": AdjacencyFamily.SPATIAL,
    "containment": AdjacencyFamily.SPATIAL,
    "overlap": AdjacencyFamily.SPATIAL,
    "path": AdjacencyFamily.SPATIAL,
    "barrier": AdjacencyFamily.SPATIAL,
    "adjacency:base": AdjacencyFamily.SPATIAL,
    "adjacency:cached": AdjacencyFamily.SPATIAL,
    "adjacency:cached-bridge": AdjacencyFamily.SPATIAL,
    "adjacency:layer:1": AdjacencyFamily.SPATIAL,
    "adjacency:layer:2": AdjacencyFamily.SPATIAL,
    "adjacency:layer:3": AdjacencyFamily.SPATIAL,
    "adjacency:layer:4": AdjacencyFamily.SPATIAL,
    "adjacency:layer:5": AdjacencyFamily.SPATIAL,
    "before": AdjacencyFamily.TEMPORAL,
    "after": AdjacencyFamily.TEMPORAL,
    "during": AdjacencyFamily.TEMPORAL,
    "recurrence": AdjacencyFamily.TEMPORAL,
    "cause": AdjacencyFamily.CAUSAL,
    "effect": AdjacencyFamily.CAUSAL,
    "enablement": AdjacencyFamily.CAUSAL,
    "inhibition": AdjacencyFamily.CAUSAL,
    "⇄": AdjacencyFamily.CAUSAL,
    "⇝": AdjacencyFamily.CAUSAL,
    "↼": AdjacencyFamily.CAUSAL,
    "seed-expansion": AdjacencyFamily.OPERATIONAL,
    "modifier:emphasis": AdjacencyFamily.COMMUNICATIVE,
    "modifier:query": AdjacencyFamily.COMMUNICATIVE,
    "modifier:left": AdjacencyFamily.COMMUNICATIVE,
    "modifier:right": AdjacencyFamily.COMMUNICATIVE,
    "modifier:close": AdjacencyFamily.COMMUNICATIVE,
    "modifier:other": AdjacencyFamily.COMMUNICATIVE,
    "self:symbol": AdjacencyFamily.AESTHETIC,
}


def classify_relation(key: str | None) -> AdjacencyFamily:
    normalized = (key or "").strip().lower()
    if not normalized:
        return AdjacencyFamily.AESTHETIC
    if normalized in RELATION_FAMILY_MAP:
        return RELATION_FAMILY_MAP[normalized]
    if normalized.startswith("adjacency:layer:"):
        return AdjacencyFamily.SPATIAL
    if normalized.startswith("adjacency:"):
        return AdjacencyFamily.SPATIAL
    if normalized.startswith("modifier:"):
        return AdjacencyFamily.COMMUNICATIVE
    if normalized == "∼":
        return AdjacencyFamily.SPATIAL
    return AdjacencyFamily.AESTHETIC

File: src/types/adjacencyFamilies.ts
export enum AdjacencyFamily {
  Spatial = 'spatial',
  Temporal = 'temporal',
  Causal = 'causal',
  Hierarchical = 'hierarchical',
  Analogical = 'analogical',
  Constraint = 'constraint',
  Value = 'value',
  Communicative = 'communicative',
  Social = 'social',
  Modal = 'modal',
  Evidential = 'evidential',
  Counterfactual = 'counterfactual',
  Operational = 'operational',
  Measurement = 'measurement',
  Aesthetic = 'aesthetic',
}

export const RELATION_FAMILY_MAP: Record<string, AdjacencyFamily> = {
  proximity: AdjacencyFamily.Spatial,
  containment: AdjacencyFamily.Spatial,
  overlap: AdjacencyFamily.Spatial,
  path: AdjacencyFamily.Spatial,
  barrier: AdjacencyFamily.Spatial,
  'adjacency:base': AdjacencyFamily.Spatial,
  'adjacency:cached': AdjacencyFamily.Spatial,
  'adjacency:cached-bridge': AdjacencyFamily.Spatial,
  'adjacency:layer:1': AdjacencyFamily.Spatial,
  'adjacency:layer:2': AdjacencyFamily.Spatial,
  'adjacency:layer:3': AdjacencyFamily.Spatial,
  'adjacency:layer:4': AdjacencyFamily.Spatial,
  'adjacency:layer:5': AdjacencyFamily.Spatial,
  'skg-base': AdjacencyFamily.Operational,
  'skg-cross-level': AdjacencyFamily.Operational,
  before: AdjacencyFamily.Temporal,
  after: AdjacencyFamily.Temporal,
  during: AdjacencyFamily.Temporal,
  recurrence: AdjacencyFamily.Temporal,
  cause: AdjacencyFamily.Causal,
  effect: AdjacencyFamily.Causal,
  enablement: AdjacencyFamily.Causal,
  inhibition: AdjacencyFamily.Causal,
  '⇄': AdjacencyFamily.Causal,
  '⇝': AdjacencyFamily.Causal,
  '↼': AdjacencyFamily.Causal,
  'seed-expansion': AdjacencyFamily.Operational,
  'modifier:emphasis': AdjacencyFamily.Communicative,
  'modifier:query': AdjacencyFamily.Communicative,
  'modifier:left': AdjacencyFamily.Communicative,
  'modifier:right': AdjacencyFamily.Communicative,
  'modifier:close': AdjacencyFamily.Communicative,
  'modifier:other': AdjacencyFamily.Communicative,
  'self:symbol': AdjacencyFamily.Aesthetic,
};

export function classifyRelation(key: string): AdjacencyFamily {
  const lower = String(key || '').toLowerCase();
  if (RELATION_FAMILY_MAP[lower]) {
    return RELATION_FAMILY_MAP[lower];
  }
  if (lower.startsWith('adjacency:layer:')) return AdjacencyFamily.Spatial;
  if (lower.startsWith('adjacency:')) return AdjacencyFamily.Spatial;
  if (lower.startsWith('modifier:')) return AdjacencyFamily.Communicative;
  if (lower === '∼') return AdjacencyFamily.Spatial;
  return AdjacencyFamily.Aesthetic;
}
